File: integration/messaging-client/message.go
package messaging

import (
	"encoding/json"
	"time"
)

// MessagePriority represents message priority levels
type MessagePriority int

const (
	PriorityLow     MessagePriority = 0
	PriorityNormal  MessagePriority = 1
	PriorityHigh    MessagePriority = 2
	PriorityCritical MessagePriority = 3
)

// Message represents a standard message
type Message struct {
	ID        string          `json:"id"`
	Timestamp time.Time       `json:"timestamp"`
	Payload   []byte          `json:"payload"`
	Topic     *string         `json:"topic,omitempty"`
	Priority  MessagePriority `json:"priority"`
}

// MessageEnvelope wraps a message with routing information
type MessageEnvelope struct {
	Message     Message           `json:"message"`
	Source      string            `json:"source"`
	Destination *string           `json:"destination,omitempty"`
	Metadata    map[string]string `json:"metadata"`
}

// MessageAck represents a message acknowledgment
type MessageAck struct {
	MessageID string    `json:"messageId"`
	Status    string    `json:"status"` // "delivered", "failed", "pending"
	Timestamp time.Time `json:"timestamp"`
}

// NewMessage creates a new message
func NewMessage(payload []byte, topic string, priority MessagePriority) Message {
	return Message{
		ID:        generateID(),
		Timestamp: time.Now(),
		Payload:   payload,
		Topic:     &topic,
		Priority:  priority,
	}
}

// NewMessageEnvelope creates a new message envelope
func NewMessageEnvelope(message Message, source string) MessageEnvelope {
	return MessageEnvelope{
		Message:  message,
		Source:   source,
		Metadata: make(map[string]string),
	}
}

// ToFrame converts a MessageEnvelope to a ProtocolFrame
func (e MessageEnvelope) ToFrame(sequenceNumber uint64) (ProtocolFrame, error) {
	payload, err := json.Marshal(e)
	if err != nil {
		return ProtocolFrame{}, err
	}
	
	return ProtocolFrame{
		Type:          FrameTypeMessage,
		Payload:       payload,
		SequenceNumber: sequenceNumber,
		Timestamp:     time.Now(),
	}, nil
}

// generateID generates a simple ID (in production, use UUID)
func generateID() string {
	return time.Now().Format("20060102150405.000000")
}

File: messaging-engine/Sources/MessagingEngine/Message.swift
import Foundation

/// Message protocol for the messaging engine
public protocol Message: Codable {
    var id: UUID { get }
    var timestamp: Date { get }
    var payload: Data { get }
}

/// Standard message implementation
public struct StandardMessage: Message {
    public let id: UUID
    public let timestamp: Date
    public let payload: Data
    public let topic: String?
    public let priority: MessagePriority
    
    public enum MessagePriority: Int, Codable {
        case low = 0
        case normal = 1
        case high = 2
        case critical = 3
    }
    
    public init(
        id: UUID = UUID(),
        timestamp: Date = Date(),
        payload: Data,
        topic: String? = nil,
        priority: MessagePriority = .normal
    ) {
        self.id = id
        self.timestamp = timestamp
        self.payload = payload
        self.topic = topic
        self.priority = priority
    }
}

/// Message envelope for network transmission
public struct MessageEnvelope: Codable {
    public let message: StandardMessage
    public let source: String
    public let destination: String?
    public let metadata: [String: String]
    
    public init(
        message: StandardMessage,
        source: String,
        destination: String? = nil,
        metadata: [String: String] = [:]
    ) {
        self.message = message
        self.source = source
        self.destination = destination
        self.metadata = metadata
    }
}

/// Message acknowledgment
public struct MessageAck: Codable {
    public let messageId: UUID
    public let status: AckStatus
    public let timestamp: Date
    
    public enum AckStatus: String, Codable {
        case delivered
        case failed
        case pending
    }
    
    public init(messageId: UUID, status: AckStatus, timestamp: Date = Date()) {
        self.messageId = messageId
        self.status = status
        self.timestamp = timestamp
    }
}

